### PR TITLE
Add support delegating requests in HttpSysServer

### PR DESCRIPTION
**Pull request by shirhatti**
Addresses https://github.com/dotnet/aspnetcore/issues/21163

### CODE REVIEW

**Comment by shirhatti** [comment]
Sample usage:

```csharp
        public void Configure(IApplicationBuilder app, IWebHostEnvironment env)
        {
            if (env.IsDevelopment())
            {
                app.UseDeveloperExceptionPage();
            }

            app.UseRouting();
            
            var delegator = app.ServerFeatures.Get<IServerDelegationPropertyFeature>();
            var delegationRule = delegator.CreateDelegationRule("DefaultAppPool", "http://*:80/");

            app.UseEndpoints(endpoints =>
            {
                endpoints.MapGet("/", context =>
                {
                    var transferFeature = context.Features.Get<IHttpSysRequestTransferFeature>();
                    transferFeature.TransferRequest(delegationRule);
                    return Task.FromResult(0);
                });
            });
        }
```

File: src/Servers/HttpSys/src/IServerDelegationPropertyFeature.cs [hunk 1]
```diff
@@ -0,0 +1,14 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
+using Microsoft.AspNetCore.Server.HttpSys;
+
+namespace Microsoft.AspNetCore.Hosting.Server.Features
+{
+    public interface IServerDelegationPropertyFeature
```
Comment (Tratcher) [inline src/Servers/HttpSys/src/IServerDelegationPropertyFeature.cs:10]:
```suggestion
    public interface IServerDelegationFeature
```

File: src/Servers/HttpSys/src/IServerDelegationPropertyFeature.cs [hunk 2]
```diff
@@ -0,0 +1,14 @@
...
+    {
+        DelegationRule CreateDelegationRule(string queueName, string uri);
```
[the remark this answers is not in the document]
Comment (Tratcher) [reply src/Servers/HttpSys/src/IServerDelegationPropertyFeature.cs:12]:
```suggestion
        DelegationRule CreateDelegationRule(string queueName, string urlPrefix);
```

File: src/Servers/HttpSys/src/NativeInterop/RequestQueue.cs [hunk 3]
```diff
@@ -16,14 +16,19 @@ internal class RequestQueue
             Marshal.SizeOf<HttpApiTypes.HTTP_BINDING_INFO>();
 
         private readonly RequestQueueMode _mode;
-        private readonly UrlGroup _urlGroup;
         private readonly ILogger _logger;
         private bool _disposed;
 
+        internal RequestQueue(UrlGroup urlGroup, string requestQueueName, RequestQueueMode mode, string Uri, ILogger logger)
```
Comment (Tratcher) [inline src/Servers/HttpSys/src/NativeInterop/RequestQueue.cs:22]:
```suggestion
        internal RequestQueue(string requestQueueName, RequestQueueMode mode, string urlPrefix, ILogger logger)
```

File: src/Servers/HttpSys/src/NativeInterop/RequestQueue.cs [hunk 4]
```diff
@@ -34,6 +39,12 @@ internal RequestQueue(UrlGroup urlGroup, string requestQueueName, RequestQueueMo
                 Created = false;
             }
 
+            if (_mode == RequestQueueMode.Receiver)
```
Comment (Tratcher) [inline src/Servers/HttpSys/src/NativeInterop/RequestQueue.cs:42]:
```suggestion
            else if (_mode == RequestQueueMode.Receiver)
```

Comment (Tratcher) [reply src/Servers/HttpSys/src/NativeInterop/RequestQueue.cs:42]: Is Receiver really a distinct mode from the others? It seems like it should be a completely different option in addition to the Create/Attach paradigm. E.g. Why can't I be delegated to when I attach to an existing queue?

File: src/Servers/HttpSys/src/NativeInterop/RequestQueue.cs [hunk 5]
```diff
@@ -54,7 +65,7 @@ internal RequestQueue(UrlGroup urlGroup, string requestQueueName, RequestQueueMo
                         out requestQueueHandle);
             }
 
-            if (flags == HttpApiTypes.HTTP_CREATE_REQUEST_QUEUE_FLAG.OpenExisting && statusCode == UnsafeNclNativeMethods.ErrorCodes.ERROR_FILE_NOT_FOUND)
+            if ((flags & HttpApiTypes.HTTP_CREATE_REQUEST_QUEUE_FLAG.OpenExisting)!=0 && statusCode == UnsafeNclNativeMethods.ErrorCodes.ERROR_FILE_NOT_FOUND)
```
Comment (Tratcher) [inline src/Servers/HttpSys/src/NativeInterop/RequestQueue.cs:68]:
```suggestion
            if ((flags.HasFlag(HttpApiTypes.HTTP_CREATE_REQUEST_QUEUE_FLAG.OpenExisting) && statusCode == UnsafeNclNativeMethods.ErrorCodes.ERROR_FILE_NOT_FOUND)
```

File: src/Servers/HttpSys/src/NativeInterop/UrlGroup.cs [hunk 6]
```diff
@@ -51,6 +68,15 @@ internal unsafe void SetMaxConnections(long maxConnections)
             SetProperty(HttpApiTypes.HTTP_SERVER_PROPERTY.HttpServerQosProperty, new IntPtr(&qosSettings), (uint)QosInfoSize);
         }
 
+        internal unsafe void SetDelegationProperty(RequestQueue queue)
```
Comment (Tratcher) [inline src/Servers/HttpSys/src/NativeInterop/UrlGroup.cs:71]:
```suggestion
        internal unsafe void SetDelegationProperty(RequestQueue destination)
```

File: src/Servers/HttpSys/src/RequestProcessing/Request.cs [hunk 7]
```diff
@@ -61,43 +61,40 @@ internal Request(RequestContext requestContext, NativeRequestContext nativeReque
...
                 var prefix = requestContext.Server.Options.UrlPrefixes.GetPrefix((int)nativeRequestContext.UrlContext);
-
-                if (originalPath.Length == prefix.PathWithoutTrailingSlash.Length)
-                {
-                    // They matched exactly except for the trailing slash.
-                    PathBase = originalPath;
-                    Path = string.Empty;
-                }
-                else
+                // Prefix may be null if the requested has been transfered to our queue
```
Comment (Tratcher) [inline src/Servers/HttpSys/src/RequestProcessing/Request.cs:76]: What will nativeRequestContext.UrlContext be in that case?

File: src/Servers/HttpSys/src/RequestProcessing/RequestContext.cs [hunk 8]
```diff
@@ -322,5 +323,36 @@ private void SetFatalResponse(int status)
             Response.ContentLength = 0;
             Dispose();
         }
+
+        internal unsafe void Transfer(DelegationRule wrapper)
```
Comment (Tratcher) [inline src/Servers/HttpSys/src/RequestProcessing/RequestContext.cs:327]:
```suggestion
        internal unsafe void Transfer(DelegationRule destination)
```

File: src/Servers/HttpSys/src/RequestProcessing/Response.cs [hunk 9]
```diff
@@ -730,6 +730,12 @@ internal void SendOpaqueUpgrade()
             }
         }
 
+        internal void Transfer()
```
Comment (Tratcher) [inline src/Servers/HttpSys/src/RequestProcessing/Response.cs:733]:
```suggestion
        internal void MarkTransfered()
```

File: src/Servers/HttpSys/src/RequestQueueMode.cs [hunk 10]
```diff
@@ -22,6 +22,10 @@ public enum RequestQueueMode
         /// Create a queue with the given name if it does not already exist, otherwise attach to the existing queue.
         /// Most configuration options do not apply when attaching to an existing queue.
         /// </summary>
-        CreateOrAttach
+        CreateOrAttach,
+        /// <summary>
+        /// Attach to an existing queue to delegate transfer requests to.
+        /// </summary>
+        Receiver
```
Comment (Tratcher) [inline src/Servers/HttpSys/src/RequestQueueMode.cs:29]: Oh, this shouldn't be public should it? It's only used in an internal API. Let's separate this to a new parameter on the request queue.

File: src/Servers/HttpSys/src/ServerDelegationPropertyFeature.cs [hunk 11]
```diff
@@ -0,0 +1,27 @@
...
+        {
+            _queue = queue;
+            _logger = logger;
+        }
+        public DelegationRule CreateDelegationRule(string queueName, string uri)
+        {
+            var wrapper = new DelegationRule(queueName, uri, _logger);
+            _queue.UrlGroup.SetDelegationProperty(wrapper.Queue);
+            return wrapper;
+        }
```
Comment (Tratcher) [inline src/Servers/HttpSys/src/ServerDelegationPropertyFeature.cs:25]:
```suggestion
        private readonly ILogger _logger;
        private readonly RequestQueue _queue;
        
        public ServerDelegationPropertyFeature(RequestQueue queue, ILogger logger)
        {
            _queue = queue;
            _logger = logger;
        }
        
        public DelegationRule CreateDelegationRule(string queueName, string uri)
        {
            var destination = new DelegationRule(queueName, uri, _logger);
            _queue.UrlGroup.SetDelegationProperty(destination.Queue);
            return destination;
        }
```

File: src/Servers/HttpSys/src/StandardFeatureCollection.cs [hunk 12]
```diff
@@ -44,6 +45,15 @@ static StandardFeatureCollection()
...
             }
+
+            try
+            {
+                if (HttpApi.HttpIsFeatureSupported(HttpApiTypes.HTTP_FEATURE_ID.HttpFeatureDelegateEx))
+                {
+                    _featureFuncLookup[typeof(IHttpSysRequestTransferFeature)] = _identityFunc;
+                }
+            }
+            catch (EntryPointNotFoundException) { }
```
Comment (Tratcher) [inline src/Servers/HttpSys/src/StandardFeatureCollection.cs:56]: You were going to move these catches to a method inside HttpApi.

**Review by Tratcher — CHANGES_REQUESTED** [review]

File: src/Servers/HttpSys/src/RequestProcessing/RequestContext.cs [hunk 13]
```diff
@@ -322,5 +323,36 @@ private void SetFatalResponse(int status)
...
+        {
+            if (Response.HasStarted)
+            {
+                throw new InvalidOperationException("This request cannot be transfered, the response has already started.");
```
Comment (shirhatti) [inline src/Servers/HttpSys/src/RequestProcessing/RequestContext.cs:331]: Need to throw if already read from RequestBody

Comment (shirhatti) [reply src/Servers/HttpSys/src/RequestProcessing/RequestContext.cs:331]: And add test for this

**Comment by shirhatti** [comment]
- Move light-up into `HttpApi.cs`. Don't have handle EntryPointNotFound in multiple places
- Don't always add the server feature, do light-up
- Add test to verify feature isn't always added
- Swap the order of MarkAsTransfered and Transfer. This ensures we can still write to the response body if transfer fails.
- Throw InvalidOperationException if we attempt to transfer after reading from RequestBody
- Add extension method to HttpContext to transfer request

File: src/Servers/HttpSys/src/RequestProcessing/RequestContext.cs [hunk 14]
```diff
@@ -322,5 +323,36 @@ private void SetFatalResponse(int status)
...
+            }
+
+            Response.Transfer();
+            var source = Server.RequestQueue;
+
+            var handle = GCHandle.Alloc(wrapper.Uri, GCHandleType.Pinned);
```
Comment (davidfowl) [inline src/Servers/HttpSys/src/RequestProcessing/RequestContext.cs:337]: I don't think I've ever seen a URI pinned to native code... Why aren't we just pinning a string here and using fixed? That would solve some issues with the codez (like if there was an exception currently calling these APIs this object will leak).

**Comment by davidfowl** [comment]
> Add extension method to HttpContext to transfer request

We shouldn't do this, it isn't common enough to move top level, even as an extension method.

File: src/Servers/HttpSys/src/RequestProcessing/Response.cs (hunk 9 above)
Comment (shirhatti) [reply src/Servers/HttpSys/src/RequestProcessing/Response.cs]: Ahh! It's `MarkTransferred` with 2 `r`s.

File: src/Servers/HttpSys/src/RequestProcessing/Request.cs (hunk 7 above)
Comment (shirhatti) [reply src/Servers/HttpSys/src/RequestProcessing/Request.cs:76]: I think this happened when using `HttpApi.DelegateRequest`. I can't repro this anymore`HttpApi.DelegateRequestEx`

File: src/Servers/HttpSys/test/FunctionalTests/DelegateTests.cs [hunk 15]
```diff
@@ -0,0 +1,187 @@
...
+
+            DelegationRule wrapper = default;
+
+            using var delegator = Utilities.CreateHttpServer(out var delegatorAddress, async httpContext =>
+            {
+                var transferFeature = httpContext.Features.Get<IHttpSysRequestTransferFeature>();
+                transferFeature.TransferRequest(wrapper);
+                await Assert.ThrowsAsync<InvalidOperationException>(async () =>
+                {
+                    await httpContext.Response.WriteAsync(_expectedResponseString);
```
Comment (Tratcher) [inline src/Servers/HttpSys/test/FunctionalTests/DelegateTests.cs:102]: Are you sure this throws? I thought you just made writes no-op. With this testing pattern you wouldn't be able to observe an assert failure here after transferring the request.

File: src/Shared/HttpSys/NativeInterop/HttpApiTypes.cs [hunk 16]
```diff
@@ -32,6 +32,8 @@ internal enum HTTP_SERVER_PROPERTY
             HttpServerListenEndpointProperty,
             HttpServerChannelBindProperty,
             HttpServerProtectionLevelProperty,
+            // Internal
```
Comment (Tratcher) [inline src/Shared/HttpSys/NativeInterop/HttpApiTypes.cs:35]:
```suggestion
```

File: src/Servers/HttpSys/src/RequestProcessing/Request.cs (hunk 7 above)
Comment (Tratcher) [reply src/Servers/HttpSys/src/RequestProcessing/Request.cs:76]: Then remove this change.

File: src/Servers/HttpSys/src/RequestProcessing/Request.cs [hunk 17]
```diff
@@ -350,6 +347,8 @@ public X509Certificate2 ClientCertificate
             }
         }
 
+        public bool IsTransferable => (HasRequestBodyStarted || RequestContext.Response.HasStarted);
```
[the remark this answers is not in the document]
Comment (shirhatti) [reply src/Servers/HttpSys/src/RequestProcessing/Request.cs:350]: Good catch! I'll add a test

File: src/Servers/HttpSys/src/NativeInterop/RequestQueue.cs [hunk 18]
```diff
@@ -16,24 +16,40 @@ internal class RequestQueue
...
                 flags = HttpApiTypes.HTTP_CREATE_REQUEST_QUEUE_FLAG.OpenExisting;
                 Created = false;
             }
 
+            if (receiver)
+            {
+                flags = HttpApiTypes.HTTP_CREATE_REQUEST_QUEUE_FLAG.OpenExisting | HttpApiTypes.HTTP_CREATE_REQUEST_QUEUE_FLAG.Delegation;
+                Created = false;
+            }
+
```
[the remark this answers is not in the document]
Comment (shirhatti) [reply src/Servers/HttpSys/src/NativeInterop/RequestQueue.cs:52]: Thank you! This is so much easier to read!

File: src/Servers/HttpSys/src/RequestProcessing/RequestContext.cs [hunk 19]
```diff
@@ -322,5 +323,45 @@ private void SetFatalResponse(int status)
...
+
+            var source = Server.RequestQueue;
+
+            uint statusCode;
+
+            fixed (char* uriPointer = destination.UrlPrefix)
+            {
+                var property = new HttpApiTypes.HTTP_DELEGATE_REQUEST_PROPERTY_INFO()
+                {
+                    ProperyId = HttpApiTypes.HTTP_DELEGATE_REQUEST_PROPERTY_ID.DelegateRequestDelegateUrlProperty,
```
[the remark this answers is not in the document]
Comment (shirhatti) [reply src/Servers/HttpSys/src/RequestProcessing/RequestContext.cs:346]: Yeah. I checked with Niranjan on this

File: src/Servers/HttpSys/test/FunctionalTests/DelegateTests.cs (hunk 15 above)
Comment (shirhatti) [reply src/Servers/HttpSys/test/FunctionalTests/DelegateTests.cs:104]: You're right. This doesn't throw.

File: src/Servers/HttpSys/src/RequestProcessing/Request.cs (hunk 7 above)
Comment (shirhatti) [reply src/Servers/HttpSys/src/RequestProcessing/Request.cs:76]: Ah got it now. If you have multiple bindings you could get an Id that doesn't correspond to an Id in UrlPrefixes.
When that happens you end up with prefix == null and fall into the else block.

I hit this issue when my backend app had two different url prefixes and delegated to the second one.

File: src/Servers/HttpSys/test/FunctionalTests/DelegateTests.cs [hunk 20]
```diff
@@ -0,0 +1,176 @@
...
+             using var receiver = Utilities.CreateHttpServer(out var receiverAddress, async httpContext =>
+            {
+                await httpContext.Response.WriteAsync(_expectedResponseString);
+            },
+            options =>
+            {
+                options.RequestQueueName = queueName;
+            });
+
+            DelegationRule wrapper = default;
```
Comment (Tratcher) [inline src/Servers/HttpSys/test/FunctionalTests/DelegateTests.cs:32]:
```suggestion
            DelegationRule destination = default;
```

File: src/Servers/HttpSys/test/FunctionalTests/DelegateTests.cs [hunk 21]
```diff
@@ -0,0 +1,176 @@
...
+
+        [ConditionalFact]
+        [DelegateSupportedCondition(true)]
+        public async Task DelegateAfterWriteToResponseBodyShouldThrowTest()
+        {
+            var queueName = Guid.NewGuid().ToString();
+            using var receiver = Utilities.CreateHttpServer(out var receiverAddress, async httpContext =>
+            {
+                await httpContext.Response.WriteAsync(_expectedResponseString);
+            },
```
Comment (Tratcher) [inline src/Servers/HttpSys/test/FunctionalTests/DelegateTests.cs:57]: Avoid the false positive if this is invoked.
```suggestion
            using var receiver = Utilities.CreateHttpServer(out var receiverAddress, httpContext =>
            {
                httpContext.Response.StatusCode = StatusCodes.Status418ImATeapot;
                return Task.CompletedTask;
            },
```

File: src/Servers/HttpSys/test/FunctionalTests/DelegateTests.cs [hunk 22]
```diff
@@ -0,0 +1,176 @@
...
+            wrapper = delegationProperty.CreateDelegationRule(queueName, receiverAddress);
+
+            var responseString = await SendRequestAsync(delegatorAddress);
+            Assert.Equal(_expectedResponseString, responseString);
+            wrapper?.Dispose();
+        }
+
+        [ConditionalFact]
+        [DelegateSupportedCondition(true)]
+        public async Task WriteToBodyAfterDelegateShouldNoOp()
```
Comment (Tratcher) [inline src/Servers/HttpSys/test/FunctionalTests/DelegateTests.cs:83]: You're not actually writing to the body anymore...

File: src/Servers/HttpSys/test/FunctionalTests/DelegateTests.cs [hunk 23]
```diff
@@ -0,0 +1,176 @@
...
+
+        [ConditionalFact]
+        [DelegateSupportedCondition(true)]
+        public async Task DelegateAfterRequestBodyReadShouldThrow()
+        {
+            var queueName = Guid.NewGuid().ToString();
+            using var receiver = Utilities.CreateHttpServer(out var receiverAddress, async httpContext =>
+            {
+                await httpContext.Response.WriteAsync(_expectedResponseString);
+            },
```
Comment (Tratcher) [inline src/Servers/HttpSys/test/FunctionalTests/DelegateTests.cs:121]:
```suggestion
            using var receiver = Utilities.CreateHttpServer(out var receiverAddress, httpContext =>
            {
                httpContext.Response.StatusCode = StatusCodes.Status418ImATeapot;
                return Task.CompletedTask;
            },
```

File: src/Shared/HttpSys/NativeInterop/HttpApiTypes.cs [hunk 24]
```diff
@@ -70,6 +71,27 @@ internal enum HTTP_TIMEOUT_TYPE
             HeaderWait,
             MinSendRate,
         }
+        internal enum HTTP_DELEGATE_REQUEST_PROPERTY_ID : uint
```
Comment (Tratcher) [inline src/Shared/HttpSys/NativeInterop/HttpApiTypes.cs:74]:
```suggestion
        }
        
        internal enum HTTP_DELEGATE_REQUEST_PROPERTY_ID : uint
```

**Review by Tratcher — APPROVED** [review]

File: src/Servers/HttpSys/test/FunctionalTests/DelegateTests.cs (hunk 22 above)
Comment (shirhatti) [reply src/Servers/HttpSys/test/FunctionalTests/DelegateTests.cs:84]: Adding back the write

File: src/Servers/HttpSys/src/FeatureContext.cs [hunk 25]
```diff
@@ -591,6 +592,8 @@ IHeaderDictionary IHttpResponseTrailersFeature.Trailers
             set => _responseTrailers = value;
         }
 
+        public bool IsTransferable => Request.IsTransferable;
```
Comment (pranavkm) [inline src/Servers/HttpSys/src/FeatureContext.cs:595]: Transferable -> `CanDelegate`

File: src/Servers/HttpSys/src/FeatureContext.cs [hunk 26]
```diff
@@ -711,5 +714,11 @@ private async Task NotifyOnCompletedAsync()
                 await actionPair.Item1(actionPair.Item2);
             }
         }
+
+        public void TransferRequest(DelegationRule destination)
```
Comment (pranavkm) [inline src/Servers/HttpSys/src/FeatureContext.cs:718]:
```suggestion
        public void DelegateRequest(DelegationRule destination)
```

File: src/Servers/HttpSys/src/IHttpSysRequestTransferFeature.cs [hunk 27]
```diff
@@ -0,0 +1,20 @@
...
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+namespace Microsoft.AspNetCore.Server.HttpSys
+{
+    public interface IHttpSysRequestTransferFeature
+    {
+        /// <summary>
+        /// Indicates if the server can transfer this request to another HttpSys request queue.
+        /// </summary>
+        bool IsTransferable { get; }
```
Comment (pranavkm) [inline src/Servers/HttpSys/src/IHttpSysRequestTransferFeature.cs:11]:
```suggestion
        bool CanDelegate { get; }
```

File: src/Servers/HttpSys/src/IHttpSysRequestTransferFeature.cs [hunk 28]
```diff
@@ -0,0 +1,20 @@
...
+
+        /// <summary>
+        /// Attempt to transfer the request to another Http.Sys request queue. The request body
+        /// must not be read nor the response started before this is invoked. Check <see cref="IsTransferable"/>
+        /// before invoking.
+        /// </summary>
+        void TransferRequest(DelegationRule destination);
```
Comment (pranavkm) [inline src/Servers/HttpSys/src/IHttpSysRequestTransferFeature.cs:18]:
```suggestion
        void DelegateRequest(DelegationRule destination);
```

File: src/Servers/HttpSys/test/FunctionalTests/DelegateTests.cs [hunk 29]
```diff
@@ -0,0 +1,179 @@
...
+                options.RequestQueueName = queueName;
+            });
+
+            DelegationRule destination = default;
+
+            using var delegator = Utilities.CreateHttpServer(out var delegatorAddress, httpContext =>
+            {
+                var delegateFeature = httpContext.Features.Get<IHttpSysRequestDelegationFeature>();
+                delegateFeature.DelegateRequest(destination);
+                return Task.FromResult(0);
```
Comment (davidfowl) [inline src/Servers/HttpSys/test/FunctionalTests/DelegateTests.cs:38]: nit: `Task.CompletedTask`

File: src/Servers/HttpSys/test/FunctionalTests/DelegateTests.cs [hunk 30]
```diff
@@ -0,0 +1,179 @@
...
+            return await client.GetStringAsync(uri);
+        }
+
+        private async Task<string> SendRequestWithBodyAsync(string uri)
+        {
+            using var client = new HttpClient();
+            var requestMessage = new HttpRequestMessage(HttpMethod.Post, uri);
+            requestMessage.Content = new StringContent("Sample request body");
+            var response = await client.SendAsync(requestMessage);
+            return await response.Content.ReadAsStringAsync();
```
Comment (Tratcher) [inline src/Servers/HttpSys/test/FunctionalTests/DelegateTests.cs:176]:
```suggestion
            using var client = new HttpClient();
            var content = new StringContent("Sample request body");
            var response = await client.PostAsync(content);
            response.EnsureSuccessStatusCode();
            return await response.Content.ReadAsStringAsync();
```

File: src/Servers/HttpSys/src/AsyncAcceptContext.cs [hunk 31]
```diff
@@ -127,7 +127,9 @@ internal uint QueueBeginGetContext()
                 statusCode = HttpApi.HttpReceiveHttpRequest(
                     Server.RequestQueue.Handle,
                     _nativeRequestContext.RequestId,
-                    (uint)HttpApiTypes.HTTP_FLAGS.HTTP_RECEIVE_REQUEST_FLAG_COPY_BODY,
```
Comment (davidfowl) [inline src/Servers/HttpSys/src/AsyncAcceptContext.cs:130]: We're making this change to allow request delegation? Will this negatively impact performance? Is so how much?

Comment (shirhatti) [reply src/Servers/HttpSys/src/AsyncAcceptContext.cs:130]: I haven't measured, but it's my understanding that in the worst case (when headers+body) happen to fit in a single TCP packet, you'll need an extra call to Http.Sys to schedule another IOCP callback.

@Tratcher ?

Comment (Tratcher) [reply src/Servers/HttpSys/src/AsyncAcceptContext.cs:130]: In the best case today you get some or all of the request body data the request struct returned from HttpReceiveHttpRequest. You still have to do another native call to verify that was the end of the body.

After this change I think you'd need to do at least two native request body reads, one for the data and one for the end of body check, though if the data were already available they could complete synchronously.

@JamesNK there are Http.Sys gRPC benchmarks with request bodies, right?

Comment (shirhatti) [reply src/Servers/HttpSys/src/AsyncAcceptContext.cs:130]: Worst case, we can always add to the `HttpSysOptions.RequestQueueMode` enum that requires you to configure your queue for delegation ahead of time and avoid opting everyone into the new behavior

Comment (Tratcher) [reply src/Servers/HttpSys/src/AsyncAcceptContext.cs:130]: Find a benchmark to run first. I don't expect there to be a difference.

Comment (shirhatti) [reply src/Servers/HttpSys/src/AsyncAcceptContext.cs:130]: Looks like there's no existing benchmark. I'll ping one of y'all offline for help in setting up a run

Comment (davidfowl) [reply src/Servers/HttpSys/src/AsyncAcceptContext.cs:130]: No existing benchmark why am surprised 😬

Comment (davidfowl) [reply src/Servers/HttpSys/src/AsyncAcceptContext.cs:130]: Why don't we add a super low level option to the API to turn this back on just in case we regress performance.

**Comment by shirhatti** [comment]
I ran a perf test with this program

```csharp
            app.Run(async context =>
            {
                await context.Request.Body.DrainAsync(CancellationToken.None);
                context.Response.StatusCode = StatusCodes.Status200OK;
            });
```

Baseline from master:

| application         |          |
| ------------------- | -------- |
| CPU Usage (%)       |       74 |
| Raw CPU Usage (%)   | 2,079.50 |
| Working Set (MB)    |      373 |
| Build Time (ms)     |    3,528 |
| Start Time (ms)     |      286 |
| Published Size (KB) |   86,882 |

| load                |           |
| ------------------- | --------- |
| CPU Usage (%)       |        36 |
| Raw CPU Usage (%)   |  1,007.63 |
| Working Set (MB)    |         7 |
| Build Time (ms)     |     3,613 |
| Start Time (ms)     |         0 |
| Published Size (KB) |    76,390 |
| First Request (ms)  |        67 |
| Requests/sec        |   565,469 |
| Requests            | 8,538,203 |
| Mean latency (ms)   |      0.62 |
| Max latency (ms)    |     32.83 |
| Bad responses       |         0 |
| Socket errors       |         0 |
| Latency 50th (ms)   |      0.30 |
| Latency 75th (ms)   |      0.71 |
| Latency 90th (ms)   |      1.21 |
| Latency 99th (ms)   |      4.04 |

My changes with `HTTP_FLAGS.NONE` (desired state)

| application         |          |
| ------------------- | -------- |
| CPU Usage (%)       |       74 |
| Raw CPU Usage (%)   | 2,074.70 |
| Working Set (MB)    |      375 |
| Build Time (ms)     |    3,515 |
| Start Time (ms)     |      298 |
| Published Size (KB) |   86,882 |

| load                |           |
| ------------------- | --------- |
| CPU Usage (%)       |        35 |
| Raw CPU Usage (%)   |    988.69 |
| Working Set (MB)    |         8 |
| Build Time (ms)     |     3,436 |
| Start Time (ms)     |         0 |
| Published Size (KB) |    76,390 |
| First Request (ms)  |        90 |
| Requests/sec        |   557,101 |
| Requests            | 8,411,906 |
| Mean latency (ms)   |      0.85 |
| Max latency (ms)    |    114.54 |
| Bad responses       |         0 |
| Socket errors       |         0 |
| Latency 50th (ms)   |      0.30 |
| Latency 75th (ms)   |      0.73 |
| Latency 90th (ms)   |      1.40 |
| Latency 99th (ms)   |      9.64 |

My changes with `HttpApiTypes.HTTP_FLAGS.HTTP_RECEIVE_REQUEST_FLAG_COPY_BODY` (control)

| application         |          |
| ------------------- | -------- |
| CPU Usage (%)       |       74 |
| Raw CPU Usage (%)   | 2,064.73 |
| Working Set (MB)    |      378 |
| Build Time (ms)     |    3,510 |
| Start Time (ms)     |      319 |
| Published Size (KB) |   86,882 |

| load                |           |
| ------------------- | --------- |
| CPU Usage (%)       |        36 |
| Raw CPU Usage (%)   |  1,000.74 |
| Working Set (MB)    |         7 |
| Build Time (ms)     |     3,538 |
| Start Time (ms)     |         0 |
| Published Size (KB) |    76,390 |
| First Request (ms)  |        90 |
| Requests/sec        |   565,653 |
| Requests            | 8,540,897 |
| Mean latency (ms)   |      0.68 |
| Max latency (ms)    |     53.49 |
| Bad responses       |         0 |
| Socket errors       |         0 |
| Latency 50th (ms)   |      0.30 |
| Latency 75th (ms)   |      0.72 |
| Latency 90th (ms)   |      1.34 |
| Latency 99th (ms)   |      5.53 |

**Comment by shirhatti** [comment]
And good measure I ran my changes with `HTTP_FLAGS.NONE` again:

| application         |          |
| ------------------- | -------- |
| CPU Usage (%)       |       74 |
| Raw CPU Usage (%)   | 2,080.62 |
| Working Set (MB)    |      373 |
| Build Time (ms)     |    3,518 |
| Start Time (ms)     |      342 |
| Published Size (KB) |   86,882 |

| load                |           |
| ------------------- | --------- |
| CPU Usage (%)       |        36 |
| Raw CPU Usage (%)   |  1,003.65 |
| Working Set (MB)    |         7 |
| Build Time (ms)     |     3,476 |
| Start Time (ms)     |         0 |
| Published Size (KB) |    76,390 |
| First Request (ms)  |        90 |
| Requests/sec        |   568,510 |
| Requests            | 8,584,097 |
| Mean latency (ms)   |      0.60 |
| Max latency (ms)    |     39.11 |
| Bad responses       |         0 |
| Socket errors       |         0 |
| Latency 50th (ms)   |      0.30 |
| Latency 75th (ms)   |      0.70 |
| Latency 90th (ms)   |      1.24 |
| Latency 99th (ms)   |      3.39 |

**Comment by shirhatti** [comment]
TL;DR: **No observable change** in perf for the identified worst case with this change.

File: src/Servers/HttpSys/src/AsyncAcceptContext.cs (hunk 31 above)
Comment (shirhatti) [reply src/Servers/HttpSys/src/AsyncAcceptContext.cs:130]: Based my on benchmark, I'm going to resolve this conversation with no new API required

**Comment by shirhatti** [comment]
@Pilchie Can I get permission to merge this change?

**Comment by Tratcher** [comment]
> @Pilchie Can I get permission to merge this change?

This is still tell mode, release/5.0 does not require additional merge permissions.
